sympy/core/numbers.py

```diff
@@ -2422,6 +2422,89 @@ def __floordiv__(self, other):
     def __rfloordiv__(self, other):
         return Integer(Integer(other).p // self.p)
 
+    def __lshift__(self, other):
+        if global_parameters.evaluate:
+            if isinstance(other, int):
+                return Integer(self.p << other)
+            elif isinstance(other, Integer):
+                return Integer(self.p << other.p)
+        return NotImplemented
+
+    def __rlshift__(self, other):
+        if global_parameters.evaluate:
+            if isinstance(other, int):
+                return Integer(other << self.p)
+            elif isinstance(other, Integer):
+                return Integer(other.p << self.p)
```

Comment:
Have you checked coverage of these methods? I think this is unreachable.

Reply:
Hmmm…now that you mention it, I _**think**_ all we need is something like this:

```python
    def __lshift__(self, other):
        if isinstance(other, numbers.Integral):
            return Integer(self.p << int(other))
        else:
            return NotImplemented

    def __rlshift__(self, other):
        if isinstance(other, numbers.Integral):
            return Integer(int(other) << self.p)
        else:
            return NotImplemented
```

Both `int` and `Integer` should be registered with `numbers.Integral`, so the above should be sufficient. Its symmetry is a _**tiny**_ bit misleading, since, as you point out, the `__r…__` methods will never be called where `other` is an `Integer`. If that hinting was important, this is redundant, but may be helpful in cultivating that understanding:

```python
    def __lshift__(self, other):
        if isinstance(other, (numbers.Integral, Integer)):
            return Integer(self.p << int(other))
        else:
            return NotImplemented

    def __rlshift__(self, other):
        if isinstance(other, numbers.Integral):
            return Integer(int(other) << self.p)
        else:
            return NotImplemented
```